### ISSUE TITLE
Incorrect result for densityratio::kliep when numerator and denominator distributions are the same

### ISSUE THREAD

**Issue by ddu10300** [thomvolker/densityratio#18]
Hello,

Thank you for developing this great package. I wanted to bring to your attention a potential issue I've encountered with the densityratio::kliep function. I used the densityratio::kliep function where both the numerator and denominator distributions are identical, however, the result deviates from 1. While the results for "ulsif" and "lhss" are close to 1.

```
library(mvtnorm)
x1 <- rmvnorm(300, mean = c(1, 1), sigma = diag(1 / 8, 2))
x2 <- x1
all.equal(x1, x2)
fit <- densityratio::kliep(
  df_numerator = x1,
  df_denominator = x2
)
pred <- predict(fit, newdata = x2)
summary(pred)
```
> summary(pred)
       V1
 Min.   :0.127
 1st Qu.:1.220
 Median :2.656
 Mean   :2.812
 3rd Qu.:3.949
 Max.   :8.029

I would greatly appreciate it if you could look into this issue. Thank you again for maintaining this great package!

**Comment by thomvolker** [comment]
Hi @ddu10300, 
Thanks for your feedback on the package! Due to a holiday break, I'm currently not in the position to deeply look into the code. However, I've noticed previously that the cross-validation scheme in 'kliep()' can be sensitive to overfitting, which seems to be the case here. Specifically, if the kernel bandwidth tends to some very small number, the cross-validation score seems to indicate better and better performance (which cannot be). You could restrict the range of bandwidth values by specifying the 'sigma' parameter manually, or set 'sigma_quantile', which by default runs from 0.05 to 0.95 (perhaps changing this to a sequence between 0.1 and 0.9 or 0.2 and 0.8 solves the issue). 

I'll look deeper into this issue upon my return!

Best,
Thom

**Comment by thomvolker** [comment]
Hi @ddu10300 ,

I had time to look into this, and it turns out that a small correction in the computation of `kliep()` had large consequences for the smallest sigma values, resulting in extremely large negative cross-validation scores. This is fixed in 128d4bb042cd29d5456aa268e98adab22c5d14a5 (see below), so I am closing the issue. Thanks again for your remark!

``` r
library(mvtnorm)
x1 <- rmvnorm(300, mean = c(1, 1), sigma = diag(1 / 8, 2))
x2 <- x1
all.equal(x1, x2)
#> [1] TRUE
fit <- densityratio::kliep(
  df_numerator = x1,
  df_denominator = x2
)
pred <- predict(fit, newdata = x2)
summary(pred)
#>        V1        
#>  Min.   :0.4885  
#>  1st Qu.:0.9613  
#>  Median :1.0213  
#>  Mean   :1.0000  
#>  3rd Qu.:1.0698  
#>  Max.   :1.1099
```

<sup>Created on 2024-10-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
